### PR TITLE
Introduce rack middleware

### PR DESCRIPTION
**Pull request by pawelniewie**
Plus additional changes:
- try to update JwtUser with details from context
- README update
- fix for user bearer token cache

### CODE REVIEW

File: lib/atlassian-jwt-authentication/filters.rb [hunk 1]
```diff
@@ -55,11 +55,14 @@ def on_add_on_installed
           if user[:username].present? && user[:display_name].present? &&
               user[:uuid].present? && user[:type].present? && user[:type] == 'user'
 
-            jwt_user = current_jwt_auth.jwt_users.where(user_key: user[:uuid]).first
-            JwtUser.create(jwt_token_id: current_jwt_auth.id,
```
Comment (bulinutza) [inline lib/atlassian-jwt-authentication/filters.rb:59]: You're not storing the jwt_token_id anymore?

File: lib/atlassian-jwt-authentication/filters.rb [hunk 2]
```diff
@@ -55,11 +55,14 @@ def on_add_on_installed
...
-                           name: user[:username],
-                           display_name: user[:display_name]) unless jwt_user
+            jwt_user = current_jwt_auth.jwt_users.find_or_initialize_by(user_key: user[:uuid]) do |jwt_user|
+              jwt_user.name = user[:username]
+              jwt_user.display_name = user[:display_name]
+            end
+
+            jwt_user.update!(
+              name: user[:username],
+              display_name: user[:display_name])
```
Comment (bulinutza) [inline lib/atlassian-jwt-authentication/filters.rb:62]: Why update the display_name here, but also set it above? I personally find the sequence of find_or_initialize_by qith a block and update! a bit confusing, I think current_jwt_auth.jwt_users.where plus create was clearer.

File: lib/atlassian-jwt-authentication/filters.rb [hunk 3]
```diff
@@ -141,108 +144,15 @@ def _verify_jwt(addon_key, consider_param = false)
...
-        head(:unauthorized)
-        return false
-      end
+      jwt_auth, jwt_user = AtlassianJwtAuthentication::Verify.verify_jwt(addon_key, jwt, request, exclude_qsh_params)
 
-      # Now verify the signature with the proper algorithm
-      begin
-        JWT.verify_signature(encoding_data['alg'], jwt_auth.shared_secret, signing_input, signature)
-      rescue Exception => e
+      if !jwt_auth
```
Comment (bulinutza) [inline lib/atlassian-jwt-authentication/filters.rb]: Pls use unless instead of if not

**Review by bulinutza — APPROVED** [review]
@pawelniewie The code looks generally good, I just had some small comments, please see below.

File: lib/atlassian-jwt-authentication/filters.rb (hunk 1 above)
Comment (pawelniewie) [reply lib/atlassian-jwt-authentication/filters.rb:59]: `jwt_token_id` will be set thanks to referencing via `current_jwt_auth.jwt_users`.

File: lib/atlassian-jwt-authentication/filters.rb (hunk 2 above)
Comment (pawelniewie) [reply lib/atlassian-jwt-authentication/filters.rb:62]: Yeah, I guess it could be simplified. Done.

**Comment by pawelniewie** [comment]
@bulinutza ok, I made small changes after your comments, this code (initializing jwt_user) will be soon gone as GDPR-ed API doesn't return user details :-(

**Pull request merged**